### PR TITLE
Issue 607 fast particle

### LINKED ISSUE #607: Fast diffusion in particles submodel
Fill out the submodel for fast diffusion in particles (i.e. ODEs not PDEs for concentration)

### PR DESCRIPTION
**Pull request by rtimms**
# Description

Adds submodel for fast diffusion in the particles. 

As a separate issue, we should allow for Broadcasts onto cell edges. This is needed here, in `constant_stefan_maxwell_diffusion` and in `isothermal` models. At the moment these all broadcast zero flux to the appropriate domain, but this ends up being evaluated on cell centers, not edges.

Fixes #607 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Key checklist:

- [x] No style issues: `$ flake8`
- [x] All tests pass: `$ python run-tests.py --unit`
- [x] The documentation builds: `$ cd docs` and then `$ make clean; make html`

You can run all three at once, using `$ python run-tests.py --quick`.

## Further checks: 

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_dfn.py [hunk 1]
```diff
@@ -86,6 +86,12 @@ def test_lumped_thermal(self):
         modeltest = tests.StandardModelTest(model, var_pts=var_pts)
         modeltest.test_all()
 
+    def test_particle_fast_diffusion(self):
+        options = {"particle": "fast diffusion"}
+        model = pybamm.lithium_ion.SPM(options)
```
Comment (martinjrobins) [inline tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_dfn.py]: Don't think this should be in the testing for DFN

File: tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_spme.py [hunk 2]
```diff
@@ -78,6 +78,12 @@ def test_thermal(self):
         modeltest = tests.StandardModelTest(model)
         modeltest.test_all()
 
+    def test_particle_fast_diffusion(self):
+        options = {"particle": "fast diffusion"}
+        model = pybamm.lithium_ion.SPM(options)
```
Comment (martinjrobins) [inline tests/integration/test_models/test_full_battery_models/test_lithium_ion/test_spme.py]: should be SPME here?

File: pybamm/models/full_battery_models/base_battery_model.py [hunk 3]
```diff
@@ -146,8 +146,9 @@ def options(self, extra_options):
             "side reactions": [],
             "interfacial surface area": "constant",
             "current collector": "uniform",
+            "particle": "Fickian diffusion",
```
Comment (martinjrobins) [inline pybamm/models/full_battery_models/base_battery_model.py]: all these options should be documented in the doctring

**Comment by rtimms** [comment]
Thanks @martinjrobins . I've started to add the options to the docstring for `BaseBatteryModel`. I wasn't sure of the correct/best way to do this, or how much detail about each of the options should be added there. I guess one option is to briefly explain what each selection you make in an option does, and another would be to give what particular submodel that option loads and then direct to the docs there for detail.

**Comment by rtimms** [comment]
Also I wasn't sure of the proper way to document things like `concatenated_rhs` in `BaseModel`. These are `None`, but then get replaced with various different thing during discretisation

**Comment by martinjrobins** [comment]
> Also I wasn't sure of the proper way to document things like `concatenated_rhs` in `BaseModel`. These are `None`, but then get replaced with various different thing during discretisation

What you have done looks good. Even though they are set to `none` in the base class, you just explain the purpose and semantic meaning of each variable so that those implementing derived classes know how they should be using them

**Comment by martinjrobins** [comment]
happy to merge when you are ready @rtimms

**Pull request merged**